### PR TITLE
[llvm] Reduce memory footprint of Debug metadata nodes

### PR DESCRIPTION
**Pull request by serge-sans-paille**
Using a combination of reordering fields and using empty SubclassData32 / SubclassData1, it's possible to improve the size of data structures used to store debug info in the IR:

Before:

DILexicalBlock: 24
DILexicalBlockFile: 24
DIModule: 24
DITemplateParameter: 24
DICommonBlock: 24
DIMacro: 24
DICompileUnit: 56
DIType: 48
DINamespace: 24
DIVariable: 24
DIGlobalVariable: 32
DILocalVariable: 32
DILabel: 24

After:

DILexicalBlock: 24
DILexicalBlockFile: 16
DIModule: 16
DITemplateParameter: 16
DICommonBlock: 16
DIMacro: 16
DICompileUnit: 48
DIType: 40
DINamespace: 16
DIVariable: 24
DIGlobalVariable: 24
DILocalVariable: 32
DILabel: 16

### CODE REVIEW

**Comment by cchen15** [comment]
Please add @stevemerr and @bwyma from the DGI team as the reviewers.

Also, you may want to consider making this change in the upstream directly.

**Review by adrian-prantl — COMMENTED** [review]
Looks like a nice improvement! Is the use of `SubclassData32` by each of the nodes documented somewhere?

**Comment by serge-sans-paille** [comment]
> Is the use of SubclassData32 by each of the nodes documented somewhere?

https://github.com/llvm/llvm-project/blob/f1fdbcba5fc0f88a481f105c2a16f56db0dd2940/llvm/include/llvm/IR/Metadata.h#L74

No documentation, unfortunately :-/

**Comment by adrian-prantl** [comment]
Let me rephrase that — could you document how SubclassData is used by the nodes you're changing in this patch? For example in the top-level doxygen comment of each DI* class you are modifying?

**Comment by serge-sans-paille** [comment]
> Let me rephrase that — could you document how SubclassData is used by the nodes you're changing in this patch? For example in the top-level doxygen comment of each DI* class you are modifying?

done!

**Comment by serge-sans-paille** [comment]
@adrian-prantl : gentle ping :-)

**Review by adrian-prantl — APPROVED** [review]
Thanks!
